File: string/valid-word-abbreviation.py
# https://leetcode.com/problems/valid-word-abbreviation/
# tags - string
class Solution:
    def validWordAbbreviation(self, word: str, abbr: str) -> bool:
        len_expanded = 0
        letter_count = ''
        for i, char in enumerate(abbr):
            if not char.isdigit():
                if len_expanded >= len(word) or word[len_expanded] != char:
                    return False
                len_expanded += 1
            else:
                if not letter_count and char == '0':
                    return False
                letter_count += char
                if i+1 == len(abbr) or not abbr[i+1].isdigit():
                    len_expanded += int(letter_count)
                    letter_count = ''
        return len_expanded == len(word)
                

File: string/valid-word-abbreviation.js
// https://leetcode.com/problems/valid-word-abbreviation/
// tags - string
/**
 * @param {string} word
 * @param {string} abbr
 * @return {boolean}
 */
var validWordAbbreviation = function (word, abbr) {
  function isDigit(char) {
    return /[0-9]/.test(char);
  }

  let j = 0;
  for (let i = 0, num = ""; i < abbr.length; i++) {
    if (isDigit(abbr[i])) {
      num += abbr[i];
      if (num === "0") return false;
      if (i === abbr.length - 1 || !isDigit(abbr[i + 1])) {
        j += Number(num);
        num = "";
      }
    } else {
      if (j >= word.length || word[j] != abbr[i]) return false;
      j++;
    }
  }

  return j === word.length;
};
